airbyte-integrations/connectors/source-zuora/source_zuora/source.py

```diff
@@ -0,0 +1,149 @@
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+
+
+from abc import ABC
+from typing import Any, Dict, Iterable, List, Mapping, MutableMapping, Tuple
+
+from airbyte_cdk import AirbyteLogger
+from airbyte_cdk.models import AirbyteStream
+from airbyte_cdk.sources import AbstractSource
+from airbyte_cdk.sources.streams import Stream
+
+from .zuora_auth import ZuoraAuthenticator
+from .zuora_client import ZoqlExportClient
+from .zuora_errors import ZOQLQueryCannotProcessObject, ZOQLQueryFieldCannotResolve
+
+
+class ZuoraStream(Stream, ABC):
+    # Define primary key
+    primary_key = "id"
+
+    def __init__(self, api: ZoqlExportClient):
+        self.api = api
+
+    def get_cursor_from_schema(self, schema: Dict) -> str:
+        """
+        Get the cursor_field from the stream's schema,
+        If the stream doesn't support 'updateddate', then we use 'createddate'.
+        """
+        return self.alt_cursor_field if self.cursor_field not in schema.keys() else self.cursor_field
+
+    def get_json_schema(self) -> Mapping[str, Any]:
+        """
+        Get the stream's schema from Zuora Object's (stream) Data types,
+        """
+        return {"properties": self.api.zuora_get_json_schema(self.name)}
+
+    def as_airbyte_stream(self) -> AirbyteStream:
+        """
+        Override this method to set the default_cursor_field to the stream schema.
+        """
+        stream = super().as_airbyte_stream()
+        stream.default_cursor_field = [self.get_cursor_from_schema(stream.json_schema["properties"])]
+        return stream
+
+    def get_stream_state(self, stream_state: Mapping[str, Any] = None) -> Mapping[str, Any]:
+        """
+        Get the state of the stream for the default cursor_field = updateddate,
+        If the stream doesn't have 'updateddate', then we use 'createddate'.
+        If stream state == None, than we use start_date by default.
+        """
+        return stream_state.get(self.cursor_field, stream_state.get(self.alt_cursor_field)) if stream_state else self.api.start_date
```

Comment:
we should use the only cursor_field here because we identified it once we extracted schema